security: TestTLSCipherRestrict failed

security.TestTLSCipherRestrict [failed](https://mesolite.cluster.engflow.com/invocations/default/b78a4eb1-ed06-4fd9-86cc-4e670b18dbe6?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc2VjdXJpdHk6c2VjdXJpdHlfdGVzdA==) on master @ [dbca69f3dc268357aefdb69776c0c720d1516203](https://github.com/cockroachdb/cockroach/commits/dbca69f3dc268357aefdb69776c0c720d1516203):


```
=== RUN   TestTLSCipherRestrict
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestTLSCipherRestrict3283140493
    test_log_scope.go:76: use -show-logs to present logs inline
    tls_test.go:268: -- test log scope end --
--- FAIL: TestTLSCipherRestrict (6.63s)
=== RUN   TestTLSCipherRestrict/no_cipher_set
    tls_test.go:217: expected wantError=false, got err=Get "https://127.0.0.1:40581/index.html": net/http: request canceled (Client.Timeout exceeded while awaiting headers), resp=
--- FAIL: TestTLSCipherRestrict/no_cipher_set (2.08s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/security-engineering @cockroachdb/server @cockroachdb/product-security
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTLSCipherRestrict.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-49911

security.TestTLSCipherRestrict [failed](https://tanzanite.cluster.engflow.com/invocations/default/0a7a328d-5c8e-4ae6-8d93-5f406041536e?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc2VjdXJpdHk6c2VjdXJpdHlfdGVzdA==) on master @ [ea65dae5f264c2e9ffb5846182e0740d487ca0b3](https://github.com/cockroachdb/cockroach/commits/ea65dae5f264c2e9ffb5846182e0740d487ca0b3):


```
=== RUN   TestTLSCipherRestrict
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestTLSCipherRestrict437123135
    test_log_scope.go:76: use -show-logs to present logs inline
    tls_test.go:268: -- test log scope end --
--- FAIL: TestTLSCipherRestrict (7.27s)
=== RUN   TestTLSCipherRestrict/no_cipher_set
    tls_test.go:217: expected wantError=false, got err=Get "https://127.0.0.1:42671/index.html": context deadline exceeded (Client.Timeout exceeded while awaiting headers), resp=
--- FAIL: TestTLSCipherRestrict/no_cipher_set (2.05s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #145812 security: TestTLSCipherRestrict failed [C-test-failure O-robot T-product-security branch-release-24.1 release-blocker]
- #145527 security: TestTLSCipherRestrict failed [C-test-failure O-robot T-db-server T-product-security branch-release-24.1.18-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTLSCipherRestrict.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/security/BUILD.bazel
+++ b/pkg/security/BUILD.bazel
@@ -93,7 +93,6 @@ go_test(
         "//pkg/server",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
-        "//pkg/testutils/skip",
         "//pkg/util/envutil",
         "//pkg/util/leaktest",
         "//pkg/util/log",

--- a/pkg/security/tls_test.go
+++ b/pkg/security/tls_test.go
@@ -13,8 +13,8 @@ import (
 	"net"
 	"net/http"
 	"strings"
+	"sync"
 	"testing"
-	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/security"
@@ -22,7 +22,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/security/username"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -124,50 +123,6 @@ func verifyX509Cert(cert *x509.Certificate, dnsName string, roots *x509.CertPool
 func TestTLSCipherRestrict(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.UnderStress(t, "http server accessing previous test's restriction fn")
-	skip.UnderRace(t)
-
-	// since the listener does not return rpc/sql/http connection errors, we
-	// need to have a separate hook to obtain and validate it.
-	type cipherErrContainer struct {
-		syncutil.Mutex
-		err net.Error
-	}
-
-	cipherErrC := &cipherErrContainer{}
-	cipherRestrictFn := security.TLSCipherRestrict
-	defer testutils.TestingHook(&security.TLSCipherRestrict, func(conn net.Conn) (err net.Error) {
-		err = cipherRestrictFn(conn)
-		cipherErrC.Lock()
-		cipherErrC.err = err
-		cipherErrC.Unlock()
-		return err
-	})()
-
-	ctx := context.Background()
-
-	// Start with a clean cipher configuration state
-	err := security.SetTLSCipherSuitesConfigured([]string{})
-	require.NoError(t, err)
-
-	s := serverutils.StartServerOnly(t, base.TestServerArgs{})
-	defer s.Stopper().Stop(ctx)
-	defer require.NoError(t, security.SetTLSCipherSuitesConfigured([]string{}))
-
-	// setup for db console tests
-	httpClient, err := s.GetUnauthenticatedHTTPClient()
-	require.NoError(t, err)
-	httpClient.Timeout = 2 * time.Second
-	defer httpClient.CloseIdleConnections()
-
-	secureClient, err := s.GetAuthenticatedHTTPClient(false, serverutils.SingleTenantSession)
-	require.NoError(t, err)
-	secureClient.Timeout = 2 * time.Second
-	defer secureClient.CloseIdleConnections()
-
-	urlsToTest := []string{"/_status/vars", "/index.html", "/"}
-	adminURLHTTPS := s.AdminURL().String()
-	adminURLHTTP := strings.Replace(adminURLHTTPS, "https", "http", 1)
 
 	tests := []struct {
 		name      string
@@ -179,9 +134,9 @@ func TestTLSCipherRestrict(t *testing.T) {
 		cipherErr string
 	}{
 		{name: "no cipher set", ciphers: []string{}, wantErr: false},
-		{name: "valid ciphers", ciphers: []string{"TLS_AES_256_GCM_SHA384", "TLS_AES_128_GCM_SHA256"},
+		{name: "valid ciphers", ciphers: []string{"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_AES_128_GCM_SHA256"},
 			wantErr: false},
-		{name: "invalid ciphers", ciphers: []string{"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"}, wantErr: true,
+		{name: "invalid ciphers", ciphers: []string{"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256"}, wantErr: true,
 			httpsErr:  []string{"\": EOF", "connect: connection refused", "read: connection reset by peer", "http: server closed idle connection"},
 			sqlErr:    "failed to connect to `host=127.0.0.1 user=root database=`: failed to receive message (unexpected EOF)",
 			rpcErr:    "initial connection heartbeat failed: grpc:",
@@ -189,38 +144,69 @@ func TestTLSCipherRestrict(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			// First ensure we're starting with no restrictions
-			err := security.SetTLSCipherSuitesConfigured([]string{})
+			// since the listener does not return rpc/sql/http connection errors, we
+			// need to have a separate hook to obtain and validate it.
+			type cipherErrContainer struct {
+				syncutil.Mutex
+				err net.Error
+			}
+
+			cipherErrC := &cipherErrContainer{}
+			cipherRestrictFn := security.TLSCipherRestrict
+			defer testutils.TestingHook(&security.TLSCipherRestrict, func(conn net.Conn) (err net.Error) {
+				err = cipherRestrictFn(conn)
+				cipherErrC.Lock()
+				cipherErrC.err = err
+				cipherErrC.Unlock()
+				return err
+			})()
+			ctx := context.Background()
+
+			s := serverutils.StartServerOnly(t, base.TestServerArgs{})
+			defer s.Stopper().Stop(ctx)
+
+			// set the custom test ciphers
+			err := security.SetTLSCipherSuitesConfigured(tt.ciphers)
 			require.NoError(t, err)
+
+			// setup for db console tests
+			httpClient, transport, err := s.ApplicationLayer().GetUnauthenticatedHTTPClientWithTransport()
+			require.NoError(t, err)
+			const httpTimeout = "Client.Timeout exceeded while awaiting headers"
+			transport.TLSClientConfig.MinVersion = tls.VersionTLS13
+			defer httpClient.CloseIdleConnections()
+
+			urlsToTest := []string{"/_status/vars", "/index.html", "/"}
+			adminURLHTTPS := s.AdminURL().String()
+			adminURLHTTP := strings.Replace(adminURLHTTPS, "https", "http", 1)
 
 			// Reset the error container before each test
 			cipherErrC.Lock()
 			cipherErrC.err = nil
 			cipherErrC.Unlock()
 
-			// Now set the custom test ciphers
-			err = security.SetTLSCipherSuitesConfigured(tt.ciphers)
-			require.NoError(t, err)
-			// unset the ciphers after test
-			defer func() { _ = security.SetTLSCipherSuitesConfigured([]string{}) }()
-
 			// test db console tls access for cipher restriction.
 			for _, u := range urlsToTest {
-				for _, client := range []http.Client{httpClient, secureClient} {
-					resp, err := client.Get(adminURLHTTP + u)
-					if (err == nil) == tt.wantErr {
-						var body []byte
+				for _, client := range []http.Client{httpClient} {
+					var wg sync.WaitGroup
+					wg.Add(1)
+					var body []byte
+					go func() {
+						defer wg.Done()
+						var resp *http.Response
+						resp, err = client.Get(adminURLHTTP + u)
 						if resp != nil && resp.Body != nil {
 							defer resp.Body.Close()
 							body, err = io.ReadAll(resp.Body)
 						}
-						t.Fatalf("expected wantError=%t, got err=%v, resp=%v", tt.wantErr, err, string(body))
+					}()
+					wg.Wait()
+					if (err == nil) == tt.wantErr {
+						if !(err != nil && strings.Contains(err.Error(), httpTimeout)) {
+							t.Fatalf("expected wantError=%t, got err=%v, resp=%v", tt.wantErr, err, string(body))
+						}
 					}
 					if tt.wantErr {
-						cipherErrC.Lock()
-						errVal := cipherErrC.err
-						cipherErrC.Unlock()
-						require.Regexp(t, tt.cipherErr, errVal.Error())
 						var errMatch bool
 						for idx := range tt.httpsErr {
 							errMatch = errMatch || strings.Contains(err.Error(), tt.httpsErr[idx])
@@ -228,6 +214,11 @@ func TestTLSCipherRestrict(t *testing.T) {
 						if !errMatch {
 							t.Fatalf("the provided error %s does not match any of the expected errors: %v", err.Error(), strings.Join(tt.httpsErr, ", "))
 						}
+						cipherErrC.Lock()
+						errVal := cipherErrC.err
+						cipherErrC.Unlock()
+						require.NotNil(t, errVal)
+						require.Regexp(t, tt.cipherErr, errVal.Error())
 					}
 				}
 			}
@@ -235,7 +226,14 @@ func TestTLSCipherRestrict(t *testing.T) {
 			// test pgx connection for root user with cert auth
 			pgURL, cleanup := s.PGUrl(t, serverutils.User(username.RootUser), serverutils.ClientCerts(true))
 			defer cleanup()
-			rootConn, err := pgx.Connect(ctx, pgURL.String())
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				_, err = pgx.Connect(ctx, pgURL.String())
+			}()
+			wg.Wait()
+
 			if (err == nil) == tt.wantErr {
 				t.Fatalf("expected wantError=%t, got err=%v", tt.wantErr, err)
 			}
@@ -245,12 +243,15 @@ func TestTLSCipherRestrict(t *testing.T) {
 				cipherErrC.Unlock()
 				require.Regexp(t, tt.cipherErr, errVal.Error())
 				require.Equal(t, tt.sqlErr, err.Error())
-			} else {
-				require.NoError(t, rootConn.Close(ctx))
 			}
 
 			// test rpc connection for root user.
-			conn, err := s.RPCClientConnE(username.RootUserName())
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				_, err = s.RPCClientConnE(username.RootUserName())
+			}()
+			wg.Wait()
 			if (err == nil) == tt.wantErr {
 				t.Fatalf("expected wantError=%t, got err=%v", tt.wantErr, err)
 			}
@@ -260,8 +261,6 @@ func TestTLSCipherRestrict(t *testing.T) {
 				cipherErrC.Unlock()
 				require.Regexp(t, tt.cipherErr, errVal.Error())
 				require.Contains(t, err.Error(), tt.rpcErr)
-			} else {
-				require.NoError(t, conn.Close()) // nolint:grpcconnclose
 			}
 		})
 	}

--- a/pkg/server/testserver_http.go
+++ b/pkg/server/testserver_http.go
@@ -91,6 +91,28 @@ func (ts *httpTestServer) GetUnauthenticatedHTTPClient() (http.Client, error) {
 	return client, nil
 }
 
+// GetUnauthenticatedHTTPClientWithTransport implements TestServerInterface.
+func (ts *httpTestServer) GetUnauthenticatedHTTPClientWithTransport() (
+	http.Client,
+	*http.Transport,
+	error,
+) {
+	client, err := ts.t.sqlServer.execCfg.RPCContext.GetHTTPClient()
+	if err != nil {
+		return client, nil, err
+	}
+	transport := client.Transport.(*http.Transport)
+	client.Transport = &tenantHeaderDecorator{
+		RoundTripper: client.Transport,
+		tenantName:   ts.t.tenantName,
+	}
+	client.Timeout = 2 * time.Second
+	if util.RaceEnabled {
+		client.Timeout = 30 * time.Second
+	}
+	return client, transport, nil
+}
+
 // GetAdminHTTPClient implements the TestServerInterface.
 func (ts *httpTestServer) GetAdminHTTPClient() (http.Client, error) {
 	httpClient, _, err := ts.GetAuthenticatedHTTPClientAndCookie(

--- a/pkg/testutils/serverutils/api.go
+++ b/pkg/testutils/serverutils/api.go
@@ -359,6 +359,12 @@ type ApplicationLayerInterface interface {
 	// with verbose method name.
 	GetUnauthenticatedHTTPClient() (http.Client, error)
 
+	// GetUnauthenticatedHTTPClientWithTransport returns an http client and its
+	// corresponding transport configured with the client TLS config required by
+	// the TestServer's configuration. Discourages implementer from using
+	// unauthenticated http connections with verbose method name.
+	GetUnauthenticatedHTTPClientWithTransport() (http.Client, *http.Transport, error)
+
 	// GetAdminHTTPClient returns an http client which has been
 	// authenticated to access Admin API methods (via a cookie).
 	// The user has admin privileges.

